File: src/php/FluxConstraint.php
<?php

declare(strict_types=1);

/**
 * FLUX Constraint Checking for PHP
 *
 * Pure PHP implementation of INT8 saturated constraint checking
 * following the FLUX constraint theory ecosystem standards.
 *
 * @package FluxConstraint
 * @version 1.0.0
 * @author FLUX Constraint Theory Ecosystem
 */

/**
 * Represents the result of a constraint check operation
 */
class FluxResult
{
    /** No violations - all constraints pass */
    public const PASS = 0;
    /** Minor violations that warrant attention */
    public const CAUTION = 1;
    /** Significant violations requiring action */
    public const WARNING = 2;
    /** Critical violations requiring immediate action */
    public const CRITICAL = 3;

    /**
     * @param int $error_mask Bitmask of violated constraints
     * @param int $severity Maximum severity level (0-3)
     * @param int $violated_lo Number of lower bound violations
     * @param int $violated_hi Number of upper bound violations
     * @param array<string, array{passed: bool, severity: int, value: int, lo: int, hi: int}> $constraint_results
     */
    public function __construct(
        public readonly int $error_mask,
        public readonly int $severity,
        public readonly int $violated_lo,
        public readonly int $violated_hi,
        public readonly array $constraint_results
    ) {}

    /**
     * Check if all constraints passed
     */
    public function isPassing(): bool
    {
        return $this->error_mask === 0;
    }

    /**
     * Check if result indicates critical failure
     */
    public function isCritical(): bool
    {
        return $this->severity >= self::CRITICAL;
    }

    /**
     * Get human-readable severity description
     */
    public function getSeverityName(): string
    {
        return match ($this->severity) {
            self::PASS => 'PASS',
            self::CAUTION => 'CAUTION',
            self::WARNING => 'WARNING',
            self::CRITICAL => 'CRITICAL',
            default => 'UNKNOWN'
        };
    }
}

/**
 * FLUX constraint checker for INT8 saturated arithmetic
 *
 * Implements constraint checking with saturated arithmetic where all values
 * are clamped to the INT8 range [-127, 127] before processing.
 */
class FluxConstraint
{
    /** Minimum INT8 value */
    public const INT8_MIN = -127;
    /** Maximum INT8 value */
    public const INT8_MAX = 127;

    /**
     * @var array<array{lo: int, hi: int, name: string, severity: int}> $constraints
     */
    private array $constraints;

    /**
     * Create a new constraint checker
     *
     * @param array<array{lo: int, hi: int, name: string, severity?: int}> $constraints
     *        Array of constraints, each with 'lo', 'hi', 'name', and optional 'severity'
     */
    public function __construct(array $constraints)
    {
        $this->constraints = [];
        foreach ($constraints as $constraint) {
            if (!isset($constraint['lo'], $constraint['hi'], $constraint['name'])) {
                throw new InvalidArgumentException('Each constraint must have lo, hi, and name');
            }

            $this->constraints[] = [
                'lo' => self::saturate((int) $constraint['lo']),
                'hi' => self::saturate((int) $constraint['hi']),
                'name' => (string) $constraint['name'],
                'severity' => (int) ($constraint['severity'] ?? FluxResult::WARNING)
            ];
        }
    }

    /**
     * Saturate a value to INT8 range [-127, 127]
     *
     * @param int $val Value to saturate
     * @return int Saturated value
     */
    public static function saturate(int $val): int
    {
        if ($val < self::INT8_MIN) {
            return self::INT8_MIN;
        }
        if ($val > self::INT8_MAX) {
            return self::INT8_MAX;
        }
        return $val;
    }

    /**
     * Check a single value against all constraints
     *
     * @param int $value Value to check (will be saturated)
     * @return FluxResult Check results
     */
    public function check(int $value): FluxResult
    {
        $saturated_value = self::saturate($value);
        $error_mask = 0;
        $max_severity = FluxResult::PASS;
        $violated_lo = 0;
        $violated_hi = 0;
        $constraint_results = [];

        foreach ($this->constraints as $i => $constraint) {
            $passed = $saturated_value >= $constraint['lo'] && $saturated_value <= $constraint['hi'];

            if (!$passed) {
                $error_mask |= (1 << $i);
                if ($saturated_value < $constraint['lo']) {
                    $violated_lo++;
                }
                if ($saturated_value > $constraint['hi']) {
                    $violated_hi++;
                }
                $max_severity = max($max_severity, $constraint['severity']);
            }

            $constraint_results[$constraint['name']] = [
                'passed' => $passed,
                'severity' => $constraint['severity'],
                'value' => $saturated_value,
                'lo' => $constraint['lo'],
                'hi' => $constraint['hi']
            ];
        }

        return new FluxResult(
            $error_mask,
            $max_severity,
            $violated_lo,
            $violated_hi,
            $constraint_results
        );
    }

    /**
     * Check multiple values against all constraints
     *
     * @param array<int> $values Values to check
     * @return array<FluxResult> Results for each value
     */
    public function checkBatch(array $values): array
    {
        $results = [];
        foreach ($values as $value) {
            $results[] = $this->check($value);
        }
        return $results;
    }

    /**
     * Create constraint checker with industry-standard presets
     *
     * @param string $name Industry preset name
     * @return self New constraint checker
     * @throws InvalidArgumentException If preset name is unknown
     */
    public static function fromIndustry(string $name): self
    {
        $presets = [
            'aviation' => [
                ['lo' => -50, 'hi' => 50, 'name' => 'altitude_deviation', 'severity' => FluxResult::CRITICAL],
                ['lo' => -20, 'hi' => 20, 'name' => 'heading_deviation', 'severity' => FluxResult::WARNING],
                ['lo' => -10, 'hi' => 10, 'name' => 'speed_deviation', 'severity' => FluxResult::WARNING]
            ],
            'medical' => [
                ['lo' => 60, 'hi' => 100, 'name' => 'heart_rate_bpm', 'severity' => FluxResult::CRITICAL],
                ['lo' => 90, 'hi' => 120, 'name' => 'systolic_bp', 'severity' => FluxResult::WARNING],
                ['lo' => 60, 'hi' => 80, 'name' => 'diastolic_bp', 'severity' => FluxResult::WARNING]
            ],
            'maritime' => [
                ['lo' => -30, 'hi' => 30, 'name' => 'course_deviation', 'severity' => FluxResult::WARNING],
                ['lo' => 0, 'hi' => 50, 'name' => 'wind_speed_knots', 'severity' => FluxResult::CAUTION],
                ['lo' => -10, 'hi' => 10, 'name' => 'depth_variance', 'severity' => FluxResult::CRITICAL]
            ],
            'automotive' => [
                ['lo' => -5, 'hi' => 5, 'name' => 'steering_angle', 'severity' => FluxResult::WARNING],
                ['lo' => 0, 'hi' => 80, 'name' => 'engine_temp_c', 'severity' => FluxResult::CRITICAL],
                ['lo' => 0, 'hi' => 120, 'name' => 'speed_kmh', 'severity' => FluxResult::WARNING]
            ],
            'industrial' => [
                ['lo' => 18, 'hi' => 25, 'name' => 'temperature_c', 'severity' => FluxResult::WARNING],
                ['lo' => 30, 'hi' => 70, 'name' => 'humidity_percent', 'severity' => FluxResult::CAUTION],
                ['lo' => -100, 'hi' => 100, 'name' => 'pressure_diff', 'severity' => FluxResult::CRITICAL]
            ],
            'financial' => [
                ['lo' => -50, 'hi' => 50, 'name' => 'price_change_percent', 'severity' => FluxResult::WARNING],
                ['lo' => 0, 'hi' => 100, 'name' => 'volatility_index', 'severity' => FluxResult::CAUTION],
                ['lo' => -20, 'hi' => 20, 'name' => 'risk_score', 'severity' => FluxResult::CRITICAL]
            ]
        ];

        if (!isset($presets[$name])) {
            throw new InvalidArgumentException("Unknown industry preset: {$name}");
        }

        return new self($presets[$name]);
    }

    /**
     * Get list of available industry presets
     *
     * @return array<string> Available preset names
     */
    public static function getAvailablePresets(): array
    {
        return ['aviation', 'medical', 'maritime', 'automotive', 'industrial', 'financial'];
    }

    /**
     * Get constraint definitions
     *
     * @return array<array{lo: int, hi: int, name: string, severity: int}> Constraint definitions
     */
    public function getConstraints(): array
    {
        return $this->constraints;
    }

    /**
     * Get total number of constraints
     */
    public function getConstraintCount(): int
    {
        return count($this->constraints);
    }
}

File: src/rust/flux_constraint.rs
//! # FLUX Constraint Engine - Rust Implementation
//!
//! A high-performance INT8 saturated constraint checking system for critical applications.
//! Provides robust boundary checking with industry-specific presets and comprehensive error reporting.
//!

#![allow(unused_imports)]
//! ## Features
//! - Zero-dependency constraint validation (no_std compatible)
//! - Industry-standard presets (aviation, medical, maritime, etc.)
//! - Batch processing capabilities
//! - Comprehensive error reporting with severity levels
//! - Memory-efficient saturated arithmetic

#![no_std]
extern crate alloc;
use alloc::vec::Vec;
use alloc::string::String;

#[cfg(test)]
#[allow(unused_imports)]
use alloc::vec;

/// INT8 minimum value for saturated arithmetic
pub const INT8_MIN: i8 = -127;

/// INT8 maximum value for saturated arithmetic
pub const INT8_MAX: i8 = 127;

/// Saturates a 32-bit integer to the valid INT8 range [-127, 127]
///
/// # Examples
/// ```
/// assert_eq!(saturate(-200), -127);
/// assert_eq!(saturate(200), 127);
/// assert_eq!(saturate(50), 50);
/// ```
pub fn saturate(val: i32) -> i8 {
    if val < INT8_MIN as i32 {
        INT8_MIN
    } else if val > INT8_MAX as i32 {
        INT8_MAX
    } else {
        val as i8
    }
}

/// A constraint definition with bounds and identifying name
#[derive(Debug, Clone, PartialEq)]
pub struct Constraint {
    /// Lower bound (inclusive)
    pub lo: i8,
    /// Upper bound (inclusive)
    pub hi: i8,
    /// Human-readable constraint name
    pub name: &'static str,
}

impl Constraint {
    /// Creates a new constraint with validation
    pub fn new(lo: i8, hi: i8, name: &'static str) -> Result<Self, FluxError> {
        if lo > hi {
            return Err(FluxError::InvalidConstraint(name));
        }
        Ok(Constraint { lo, hi, name })
    }
}

/// Severity levels for constraint violations
#[derive(Debug, Clone, Copy, PartialEq)]
#[repr(u8)]
pub enum Severity {
    Pass = 0,
    Caution = 1,
    Warning = 2,
    Critical = 3,
}

/// Result of constraint checking with detailed violation information
#[derive(Debug, Clone, PartialEq)]
pub struct FluxResult {
    /// Bitmask of which constraints failed (bit i = constraint i)
    pub error_mask: u8,
    /// Maximum severity level encountered
    pub severity: u8,
    /// Count of lower bound violations
    pub violated_lo: u8,
    /// Count of upper bound violations
    pub violated_hi: u8,
}

impl FluxResult {
    /// Creates a new result indicating all constraints passed
    pub fn pass() -> Self {
        FluxResult {
            error_mask: 0,
            severity: Severity::Pass as u8,
            violated_lo: 0,
            violated_hi: 0,
        }
    }

    /// Checks if all constraints passed
    pub fn is_pass(&self) -> bool {
        self.error_mask == 0
    }

    /// Gets the severity as an enum
    pub fn get_severity(&self) -> Severity {
        match self.severity {
            0 => Severity::Pass,
            1 => Severity::Caution,
            2 => Severity::Warning,
            _ => Severity::Critical,
        }
    }
}

/// Error types for the FLUX system
#[derive(Debug, Clone, PartialEq)]
pub enum FluxError {
    InvalidConstraint(&'static str),
    TooManyConstraints,
    PresetNotFound(String),
}

/// High-performance constraint checker with batch processing capabilities
pub struct FluxChecker {
    constraints: Vec<Constraint>,
}

impl FluxChecker {
    /// Creates a new checker with the given constraints
    ///
    /// # Arguments
    /// * `constraints` - Vector of constraints to enforce
    ///
    /// # Errors
    /// Returns `FluxError::TooManyConstraints` if more than 8 constraints are provided
    /// (limited by error_mask being u8)
    pub fn new(constraints: Vec<Constraint>) -> Result<Self, FluxError> {
        if constraints.len() > 8 {
            return Err(FluxError::TooManyConstraints);
        }
        Ok(FluxChecker { constraints })
    }

    /// Checks a single value against all constraints
    ///
    /// # Arguments
    /// * `value` - The INT8 value to check
    ///
    /// # Returns
    /// FluxResult with detailed violation information
    pub fn check(&self, value: i8) -> FluxResult {
        let mut result = FluxResult::pass();

        for (i, constraint) in self.constraints.iter().enumerate() {
            if value < constraint.lo {
                result.error_mask |= 1 << i;
                result.violated_lo += 1;
                result.severity = result.severity.max(self.calculate_severity(constraint, value));
            } else if value > constraint.hi {
                result.error_mask |= 1 << i;
                result.violated_hi += 1;
                result.severity = result.severity.max(self.calculate_severity(constraint, value));
            }
        }

        result
    }

    /// Checks multiple values in batch for improved performance
    ///
    /// # Arguments
    /// * `values` - Slice of INT8 values to check
    ///
    /// # Returns
    /// Vector of FluxResult, one for each input value
    pub fn check_batch(&self, values: &[i8]) -> Vec<FluxResult> {
        values.iter().map(|&value| self.check(value)).collect()
    }

    /// Creates a checker from an industry preset
    ///
    /// # Arguments
    /// * `name` - Name of the preset ("aviation", "medical", "maritime", "automotive", "energy", "nuclear")
    ///
    /// # Returns
    /// Configured FluxChecker or error if preset not found
    pub fn from_preset(name: &str) -> Result<Self, FluxError> {
        let constraints = match name {
            "aviation" => AVIATION_CONSTRAINTS.to_vec(),
            "medical" => MEDICAL_CONSTRAINTS.to_vec(),
            "maritime" => MARITIME_CONSTRAINTS.to_vec(),
            "automotive" => AUTOMOTIVE_CONSTRAINTS.to_vec(),
            "energy" => ENERGY_CONSTRAINTS.to_vec(),
            "nuclear" => NUCLEAR_CONSTRAINTS.to_vec(),
            _ => return Err(FluxError::PresetNotFound(String::from(name))),
        };

        FluxChecker::new(constraints)
    }

    /// Calculates severity based on how far the value is from constraint bounds
    fn calculate_severity(&self, constraint: &Constraint, value: i8) -> u8 {
        let range = (constraint.hi as i16) - (constraint.lo as i16);
        let violation_distance = if value < constraint.lo {
            (constraint.lo as i16) - (value as i16)
        } else {
            (value as i16) - (constraint.hi as i16)
        };

        if range == 0 || violation_distance >= range {
            Severity::Critical as u8
        } else if violation_distance >= range / 2 {
            Severity::Warning as u8
        } else {
            Severity::Caution as u8
        }
    }
}

// Industry-specific constraint presets
const AVIATION_CONSTRAINTS: &[Constraint] = &[
    Constraint { lo: -50, hi: 85, name: "temperature_celsius" },
    Constraint { lo: 0, hi: 100, name: "cabin_pressure_percent" },
    Constraint { lo: -90, hi: 90, name: "pitch_degrees" },
    Constraint { lo: -127, hi: 127, name: "roll_degrees_scaled" },
];

const MEDICAL_CONSTRAINTS: &[Constraint] = &[
    Constraint { lo: 60, hi: 100, name: "heart_rate_bpm_normalized" },
    Constraint { lo: 12, hi: 20, name: "respiratory_rate" },
    Constraint { lo: 36, hi: 38, name: "body_temp_celsius" },
    Constraint { lo: 90, hi: 100, name: "blood_oxygen_percent" },
];

const MARITIME_CONSTRAINTS: &[Constraint] = &[
    Constraint { lo: -40, hi: 60, name: "sea_temp_celsius" },
    Constraint { lo: 0, hi: 127, name: "wave_height_decimeters" },
    Constraint { lo: 0, hi: 120, name: "wind_speed_knots" },
    Constraint { lo: -50, hi: 50, name: "pressure_deviation_hpa" },
];

const AUTOMOTIVE_CONSTRAINTS: &[Constraint] = &[
    Constraint { lo: -40, hi: 125, name: "engine_temp_celsius" },
    Constraint { lo: 0, hi: 127, name: "vehicle_speed_mph" },
    Constraint { lo: 0, hi: 100, name: "fuel_level_percent" },
    Constraint { lo: 11, hi: 15, name: "battery_voltage" },
];

const ENERGY_CONSTRAINTS: &[Constraint] = &[
    Constraint { lo: 110, hi: 125, name: "grid_voltage_normalized" },
    Constraint { lo: 49, hi: 51, name: "frequency_hz" },
    Constraint { lo: 0, hi: 100, name: "load_factor_percent" },
    Constraint { lo: -20, hi: 80, name: "ambient_temp_celsius" },
];

const NUCLEAR_CONSTRAINTS: &[Constraint] = &[
    Constraint { lo: -40, hi: 40, name: "reactor_temp_deviation_celsius" },
    Constraint { lo: 0, hi: 127, name: "radiation_level_normalized" },
    Constraint { lo: 14, hi: 16, name: "pressure_mpa" },
    Constraint { lo: 0, hi: 100, name: "control_rod_position_percent" },
];

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::vec;

    #[test]
    fn test_saturate_boundaries() {
        assert_eq!(saturate(-128), -127);
        assert_eq!(saturate(-127), -127);
        assert_eq!(saturate(0), 0);
        assert_eq!(saturate(127), 127);
        assert_eq!(saturate(128), 127);
        assert_eq!(saturate(1000), 127);
        assert_eq!(saturate(-1000), -127);
    }

    #[test]
    fn test_constraint_creation() {
        let constraint = Constraint::new(-10, 10, "test").unwrap();
        assert_eq!(constraint.lo, -10);
        assert_eq!(constraint.hi, 10);
        assert_eq!(constraint.name, "test");

        // Invalid constraint (lo > hi)
        assert!(Constraint::new(10, -10, "invalid").is_err());
    }

    #[test]
    fn test_single_constraint_pass() {
        let constraints = vec![
            Constraint::new(-10, 10, "range_test").unwrap()
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        let result = checker.check(5);
        assert!(result.is_pass());
        assert_eq!(result.error_mask, 0);
        assert_eq!(result.severity, Severity::Pass as u8);
        assert_eq!(result.violated_lo, 0);
        assert_eq!(result.violated_hi, 0);
    }

    #[test]
    fn test_single_constraint_fail_low() {
        let constraints = vec![
            Constraint::new(-10, 10, "range_test").unwrap()
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        let result = checker.check(-15);
        assert!(!result.is_pass());
        assert_eq!(result.error_mask, 1);
        assert_eq!(result.violated_lo, 1);
        assert_eq!(result.violated_hi, 0);
        assert!(result.severity > 0);
    }

    #[test]
    fn test_single_constraint_fail_high() {
        let constraints = vec![
            Constraint::new(-10, 10, "range_test").unwrap()
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        let result = checker.check(15);
        assert!(!result.is_pass());
        assert_eq!(result.error_mask, 1);
        assert_eq!(result.violated_lo, 0);
        assert_eq!(result.violated_hi, 1);
        assert!(result.severity > 0);
    }

    #[test]
    fn test_multi_constraint_mixed_results() {
        let constraints = vec![
            Constraint::new(-10, 10, "range1").unwrap(),
            Constraint::new(0, 20, "range2").unwrap(),
            Constraint::new(-5, 5, "range3").unwrap(),
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        let result = checker.check(8);
        // Should pass range1 (8 in [-10,10]) and range2 (8 in [0,20])
        // Should fail range3 (8 not in [-5,5])
        assert!(!result.is_pass());
        assert_eq!(result.error_mask, 0b100); // Third constraint failed
        assert_eq!(result.violated_hi, 1);
        assert_eq!(result.violated_lo, 0);
    }

    #[test]
    fn test_severity_calculation() {
        let constraints = vec![
            Constraint::new(-10, 10, "narrow_range").unwrap(),
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        // Small violation should be caution
        let result = checker.check(12);
        assert_eq!(result.get_severity(), Severity::Caution);

        // Large violation should be critical
        let result = checker.check(50);
        assert_eq!(result.get_severity(), Severity::Critical);
    }

    #[test]
    fn test_batch_checking() {
        let constraints = vec![
            Constraint::new(-10, 10, "range_test").unwrap()
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        let values = vec![5, -15, 15, 0];
        let results = checker.check_batch(&values);

        assert_eq!(results.len(), 4);
        assert!(results[0].is_pass()); // 5 is in range
        assert!(!results[1].is_pass()); // -15 is below range
        assert!(!results[2].is_pass()); // 15 is above range
        assert!(results[3].is_pass()); // 0 is in range
    }

    #[test]
    fn test_preset_loading() {
        let aviation = FluxChecker::from_preset("aviation").unwrap();
        assert_eq!(aviation.constraints.len(), 4);

        let medical = FluxChecker::from_preset("medical").unwrap();
        assert_eq!(medical.constraints.len(), 4);

        let maritime = FluxChecker::from_preset("maritime").unwrap();
        assert_eq!(maritime.constraints.len(), 4);

        let automotive = FluxChecker::from_preset("automotive").unwrap();
        assert_eq!(automotive.constraints.len(), 4);

        let energy = FluxChecker::from_preset("energy").unwrap();
        assert_eq!(energy.constraints.len(), 4);

        let nuclear = FluxChecker::from_preset("nuclear").unwrap();
        assert_eq!(nuclear.constraints.len(), 4);

        // Unknown preset should fail
        assert!(FluxChecker::from_preset("unknown").is_err());
    }

    #[test]
    fn test_overflow_protection() {
        // Test that we can't create more than 8 constraints
        let mut constraints = Vec::new();
        for _i in 0..10 {
            constraints.push(Constraint::new(-1, 1, "test").unwrap());
        }
        assert!(FluxChecker::new(constraints).is_err());
    }

    #[test]
    fn test_edge_cases() {
        let constraints = vec![
            Constraint::new(INT8_MIN, INT8_MAX, "full_range").unwrap()
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        // Min and max values should pass
        assert!(checker.check(INT8_MIN).is_pass());
        assert!(checker.check(INT8_MAX).is_pass());

        // Test with empty constraints
        let empty_checker = FluxChecker::new(vec![]).unwrap();
        assert!(empty_checker.check(0).is_pass());
    }

    #[test]
    fn test_all_pass_scenario() {
        let constraints = vec![
            Constraint::new(-50, 50, "wide_range1").unwrap(),
            Constraint::new(-100, 100, "wide_range2").unwrap(),
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        let result = checker.check(25);
        assert!(result.is_pass());
        assert_eq!(result.error_mask, 0);
        assert_eq!(result.violated_lo, 0);
        assert_eq!(result.violated_hi, 0);
    }

    #[test]
    fn test_all_fail_scenario() {
        let constraints = vec![
            Constraint::new(10, 20, "high_range1").unwrap(),
            Constraint::new(15, 25, "high_range2").unwrap(),
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        let result = checker.check(5);
        assert!(!result.is_pass());
        assert_eq!(result.error_mask, 0b11); // Both constraints failed
        assert_eq!(result.violated_lo, 2);
        assert_eq!(result.violated_hi, 0);
    }
}

/// Simple benchmarking module using inline timing
#[cfg(test)]
mod benchmarks {
    use super::*;
    use alloc::vec;

    /// Get current time in nanoseconds (mock implementation for no_std)
    fn get_time_ns() -> u64 {
        // In a real no_std environment, you'd use a hardware timer
        // For testing purposes, we'll use a simple counter
        static mut COUNTER: u64 = 0;
        unsafe {
            COUNTER += 1000; // Simulate 1us per call
            COUNTER
        }
    }

    #[test]
    fn benchmark_single_checks() {
        let constraints = vec![
            Constraint::new(-50, 50, "temp").unwrap(),
            Constraint::new(0, 100, "pressure").unwrap(),
            Constraint::new(-90, 90, "angle").unwrap(),
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        let iterations = 10000;
        let start_time = get_time_ns();

        for i in 0..iterations {
            let value = ((i % 201) - 100) as i8; // Values from -100 to 100
            let _ = checker.check(value);
        }

        let end_time = get_time_ns();
        let duration_ns = end_time - start_time;
        let checks_per_sec = (iterations as f64 * 1_000_000_000.0) / duration_ns as f64;

        // Store results in variables to verify calculations
        let _single_checks_per_sec = checks_per_sec;
        let _single_duration_ns = duration_ns;

        // Verify we achieved reasonable performance (mock timing always passes)
        assert!(checks_per_sec > 0.0);
    }

    #[test]
    fn benchmark_batch_checks() {
        let constraints = vec![
            Constraint::new(-50, 50, "temp").unwrap(),
            Constraint::new(0, 100, "pressure").unwrap(),
            Constraint::new(-90, 90, "angle").unwrap(),
        ];
        let checker = FluxChecker::new(constraints).unwrap();

        let batch_size = 1000;
        let values: Vec<i8> = (0..batch_size).map(|i| ((i % 201) - 100) as i8).collect();
        let iterations = 100;

        let start_time = get_time_ns();

        for _ in 0..iterations {
            let _ = checker.check_batch(&values);
        }

        let end_time = get_time_ns();
        let duration_ns = end_time - start_time;
        let total_checks = iterations * batch_size;
        let checks_per_sec = (total_checks as f64 * 1_000_000_000.0) / duration_ns as f64;

        // Store results in variables for verification
        let _batch_checks_per_sec = checks_per_sec;
        let _batch_size_result = batch_size;
        let _iterations_result = iterations;
        let _duration_result = duration_ns;

        // Verify we achieved reasonable performance (mock timing always passes)
        assert!(checks_per_sec > 0.0);
    }

    #[test]
    fn benchmark_preset_loading() {
        let presets = ["aviation", "medical", "maritime", "automotive", "energy", "nuclear"];

        let start_time = get_time_ns();

        for preset in &presets {
            for _ in 0..1000 {
                let _ = FluxChecker::from_preset(preset).unwrap();
            }
        }

        let end_time = get_time_ns();
        let duration_ns = end_time - start_time;
        let loads_per_sec = (presets.len() * 1000) as f64 * 1_000_000_000.0 / duration_ns as f64;

        // Store results for verification
        let _preset_loads_per_sec = loads_per_sec;
        let _total_presets_loaded = presets.len() * 1000;
        let _preset_duration = duration_ns;

        assert!(loads_per_sec > 0.0);
    }
}